presto-tests/src/test/java/io/prestosql/tests/TestProcedureCreation.java

```diff
@@ -0,0 +1,183 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowableOfType;
+
+@Test(singleThreaded = true)
+public class TestProcedureCreation
+{
+    @Test
+    public void shouldThrowExceptionWhenOptionalArgumentIsNotLast()
+    {
+        assertThrows(
+                IllegalArgumentException.class,
+                "Optional arguments should follow required ones",
+                () -> createTestProcedure(ImmutableList.of(
+                    new Procedure.Argument("name", VARCHAR, false, null),
+                    new Procedure.Argument("name2", VARCHAR, true, null))));
+
+        assertThrows(
+                IllegalArgumentException.class,
+                "Optional arguments should follow required ones",
+                () -> createTestProcedure(ImmutableList.of(
+                        new Procedure.Argument("name", VARCHAR, true, null),
+                        new Procedure.Argument("name2", VARCHAR, true, null),
+                        new Procedure.Argument("name3", VARCHAR, true, null),
+                        new Procedure.Argument("name4", VARCHAR, false, null),
+                        new Procedure.Argument("name5", VARCHAR, true, null))));
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenArgumentNameRepeates()
+    {
+        assertThrows(
+                IllegalArgumentException.class,
+                "Duplicate argument name: 'name'",
+                () -> createTestProcedure(ImmutableList.of(
+                        new Procedure.Argument("name", VARCHAR, false, null),
+                        new Procedure.Argument("name", VARCHAR, true, null))));
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenProcedureIsNonVoid()
+    {
+        assertThrows(
+                IllegalArgumentException.class,
+                "Method must return void",
+                () -> new Procedure(
+                        "schema",
+                        "name",
+                        ImmutableList.of(),
+                        methodHandle(Procedures.class, "funWithoutArguments")));
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenMethodHandleIsNull()
+    {
+        assertThrows(
+                NullPointerException.class,
+                "methodHandle is null",
+                () -> new Procedure(
+                        "schema",
+                        "name",
+                        ImmutableList.of(),
+                        null));
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenMethodHandleHasVarargs()
+    {
+        assertThrows(
+                IllegalArgumentException.class,
+                "Method must have fixed arity",
+                () -> new Procedure(
+                        "schema",
+                        "name",
+                        ImmutableList.of(),
+                        methodHandle(Procedures.class, "funWithVarargs", String[].class)));
```

Comment:
Any reason not to use `org.assertj.core.api.Assertions#assertThatThrownBy` directly?
it wouldn't be longer, yet no help needed:

```
assertThatThrownBy(() -> new Procedure(
        "schema",
        "name",
        ImmutableList.of(),
        methodHandle(Procedures.class, "funWithVarargs", String[].class)))
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessage("Method must have fixed arity");
```